python/ir/bindings.cpp

```diff
@@ -0,0 +1,133 @@
+        .value("u32", BaseType::UINT32)
+        .value("u64", BaseType::UINT64)
+        .value("f32", BaseType::FLOAT32)
+        .value("f64", BaseType::FLOAT64)
+        .value("struct", BaseType::STRUCT)
+        .value("ptr", BaseType::PTR);
+
+    py::class_<DataType>(m, "DataType")
+        .def(py::init<BaseType, vector<DataType>, size_t>(), py::arg("btype"),
+             py::arg("dtypes") = vector<DataType>{}, py::arg("size") = 0)
+        .def("str", &DataType::str);
+
+    /* StmtNode and Derived Structures Declarations
+     */
+    py::class_<StmtNode, Stmt>(m, "stmt");
+    py::class_<ExprNode, Expr, StmtNode>(m, "expr");
+
+    /* Symbol Definition */
+    py::class_<SymNode, Sym, ExprNode>(m, "sym")
+        .def(py::init<string, DataType>())
+        .def(py::init<string, Expr>());
+
+    /* Constant Expressions */
+    REGISTER_CLASS(Const, ExprNode, m, "const", DataType, double)
+
+    /* Loop ops */
+    REGISTER_CLASS(IsValid, ExprNode, m, "is_valid", Expr, Expr, size_t)
+    REGISTER_CLASS(SetValid, ExprNode, m, "set_valid", Expr, Expr, Expr, size_t)
+    REGISTER_CLASS(FetchDataPtr, ExprNode, m, "fetch", Expr, Expr, size_t)
+    REGISTER_CLASS(Alloc, ExprNode, m, "alloc", DataType, Expr)
+    REGISTER_CLASS(Load, ExprNode, m, "load", Expr)
+    REGISTER_CLASS(Store, StmtNode, m, "store", Expr, Expr)
+    REGISTER_CLASS(Loop, ExprNode, m, "loop", Expr)
+
+    /* Statements */
+    REGISTER_CLASS(Func, StmtNode, m, "func", string, Expr, vector<Sym>,
+                   SymTable)
+    REGISTER_CLASS(Stmts, StmtNode, m, "stmts", vector<Stmt>)
+    REGISTER_CLASS(IfElse, StmtNode, m, "ifelse", Expr, Stmt, Stmt)
+    REGISTER_CLASS(NoOp, StmtNode, m, "noop")
+
+    /* Math Operators for Nary Expressions */
+    py::enum_<MathOp>(m, "MathOp")
+        .value("add", MathOp::ADD)
+        .value("sub", MathOp::SUB)
+        .value("mul", MathOp::MUL)
+        .value("div", MathOp::DIV)
+        .value("max", MathOp::MAX)
+        .value("min", MathOp::MIN)
+        .value("mod", MathOp::MOD)
+        .value("sqrt", MathOp::SQRT)
+        .value("pow", MathOp::POW)
+        .value("abs", MathOp::ABS)
+        .value("neg", MathOp::NEG)
+        .value("ceil", MathOp::CEIL)
+        .value("floor", MathOp::FLOOR)
+        .value("lt", MathOp::LT)
+        .value("lte", MathOp::LTE)
+        .value("gt", MathOp::GT)
+        .value("gte", MathOp::GTE)
+        .value("eq", MathOp::EQ)
+        .value("not", MathOp::NOT)
+        .value("and", MathOp::AND)
+        .value("or", MathOp::OR);
+
+    /* Nary Expressions */
+    REGISTER_CLASS(NaryExpr, ExprNode, m, "nary_expr", DataType, MathOp,
+                   vector<Expr>)
+    REGISTER_CLASS(UnaryExpr, NaryExpr, m, "unary_expr", DataType, MathOp, Expr)
+    REGISTER_CLASS(BinaryExpr, NaryExpr, m, "binary_expr", DataType, MathOp,
+                   Expr, Expr)
+
+    /* Logical Expressions */
+    REGISTER_CLASS(Equals, BinaryExpr, m, "equals", Expr, Expr)
```

Comment:
add the definitions of all the expressions and stmts. basically everything in the ir/ directory

Reply:
done.